pkg/operation/etcdencryption/encryptionconfiguration.go

```diff
@@ -0,0 +1,218 @@
+	if !sets.NewString(ec.Resources[0].Resources...).Equal(sets.NewString(ecEncryptedResources...)) {
+		return false, fmt.Errorf("list of encrypted resources (%v) of the EncryptionConfiguration does not match the expected list (%v)", ec.Resources[0].Resources, ecEncryptedResources)
+	}
+	// Check encryption providers
+	ecEncryptionProviderLenA := len(ec.Resources[0].Providers)
+	if ecEncryptionProviderLenA != ecEncryptionProviderLenE {
+		return false, fmt.Errorf("number of encryption providers (%v) of the EncryptionConfiguration does not match the expected number (%v)", ecEncryptionProviderLenA, ecEncryptionProviderLenE)
+	}
+	// Encryption provider aescbc and identity in any order are ok
+	if !(((ec.Resources[0].Providers[0].Identity != nil) && (ec.Resources[0].Providers[1].AESCBC != nil)) ||
+		((ec.Resources[0].Providers[0].AESCBC != nil) && (ec.Resources[0].Providers[1].Identity != nil))) {
+		return false, fmt.Errorf("unexpected encryption providers of the EncryptionConfiguration found. Expected are exactly two providers 'identity' and 'aescbc'")
+	}
+	var aesConfig *apiserverconfigv1.AESConfiguration
+	if ec.Resources[0].Providers[0].AESCBC != nil {
+		aesConfig = ec.Resources[0].Providers[0].AESCBC
+	} else {
+		aesConfig = ec.Resources[0].Providers[1].AESCBC
+	}
+	ecEncryptionProviderAESCBCKeyLenA := len(aesConfig.Keys)
+	if ecEncryptionProviderAESCBCKeyLenA < ecEncryptionProviderAESCBCMinKeyLenE {
+		return false, fmt.Errorf("unexpected number of keys in encryption provider aescbc of the EncryptionConfiguration found. Expected are at least %v key(s)", ecEncryptionProviderAESCBCMinKeyLenE)
+	}
+	keyNameMap := make(map[string]bool)
+	for _, key := range aesConfig.Keys {
+		ok, err := isKeyConsistent(key)
+		if (err != nil) || (!ok) {
+			return false, fmt.Errorf("inconsistent key in encryption provider aescbc of the EncryptionConfiguration found")
+		}
+		_, ok = keyNameMap[key.Name]
+		if ok {
+			return false, fmt.Errorf("two or more keys (%v) with same timestamp found in encryption provider aescbc of the EncryptionConfiguration", key.Name)
+		}
+		keyNameMap[key.Name] = true
+	}
+	return true, nil
+}
+
+// Equals checks whether the provided encryption configurations are equal.
+func Equals(ec1 *apiserverconfigv1.EncryptionConfiguration, ec2 *apiserverconfigv1.EncryptionConfiguration) (bool, error) {
+	ec1YamlBytes, err := ToYAML(ec1)
+	if err != nil {
+		return false, fmt.Errorf("error when converting EncryptionConfiguration to yaml: %v", err)
+	}
+	ec1YamlString := string(ec1YamlBytes)
+	ec2YamlBytes, err := ToYAML(ec2)
+	if err != nil {
+		return false, fmt.Errorf("error when converting EncryptionConfiguration to yaml: %v", err)
+	}
+	ec2YamlString := string(ec2YamlBytes)
+	return ec1YamlString == ec2YamlString, nil
+}
+
+// IsActive checks whether the EncryptionConfiguration is active, i.e. whether the provider
+// identity is NOT the first in the list of providers.
+func IsActive(ec *apiserverconfigv1.EncryptionConfiguration) bool {
```

Comment:
Name is a bit misleading I guess - you rather check whether a certain provider at index 0 is there. What if there are multiple providers / different indexes? Please fix the name and the implementation of this method.

Reply:
I check that the EncryptionConfiguration is consistent (or Validate, whichever name you prefer :) ) and part of this is to check that there are exactly two providers (aescbc and identity) in arbitrary order. 

Only once the EncryptionConfiguration is validated isActive is called, and thus I do not have to check - again - that there may be more providers/other providers etc.